library/Icingadb/Data/PivotTable.php

```diff
@@ -0,0 +1,410 @@
+            }
+        }
+
+        $query = $this->queryXAxis();
+
+        $query->limit($limit);
+
+        $query->offset($page > 0 ? ($page - 1) * $limit : 0);
+
+        return $query;
+    }
+
+    /**
+     * Return a Paginatable for the y-axis query
+     *
+     * $limit and $page are taken from the current request if not given.
+     *
+     * @param   int     $limit  The maximum amount of entries to fetch
+     * @param   int     $page   The page to set as current one
+     *
+     * @return  Paginatable
+     */
+    public function paginateYAxis($limit = null, $page = null)
+    {
+        if ($limit === null || $page === null) {
+            if ($limit === null) {
+                $limit = $this->getPaginationParameter('y', 'limit', 20);
+            }
+
+            if ($page === null) {
+                $page = $this->getPaginationParameter('y', 'page', 1);
+            }
+        }
+
+        $query = $this->queryYAxis();
+        $query->limit($limit);
+        $query->offset($page > 0 ? ($page - 1) * $limit : 0);
+
+        return $query;
+    }
+
+    /**
+     * Return the pivot table as an array of pivot data and pivot header
+     *
+     * @return array
+     */
+    public function toArray()
+    {
+        if (
+            ($this->xAxisFilter === null && $this->yAxisFilter === null)
+            || ($this->xAxisFilter !== null && $this->yAxisFilter !== null)
+        ) {
+            $xAxis = $this->getAxisKeys($this->queryXAxis(), $this->xAxisColumn);
+            $xAxisKeys = array_keys($xAxis);
+            $yAxis = $this->getAxisKeys($this->queryYAxis(), $this->yAxisColumn);
+            $yAxisKeys = array_keys($yAxis);
+        } else {
+            if ($this->xAxisFilter !== null) {
+                $xAxis = $this->getAxisKeys($this->queryXAxis(), $this->xAxisColumn);
+                $xAxisKeys = array_keys($xAxis);
+                $yQuery = $this->queryYAxis();
+                $yQuery->getSelectBase()->where($this->xAxisColumn, $xAxisKeys);
+                $yAxis = $this->getAxisKeys($yQuery, $this->yAxisColumn);
+                $yAxisKeys = array_keys($yAxis);
+            } else { // $this->yAxisFilter !== null
+                $yAxis = $this->getAxisKeys($this->queryYAxis(), $this->yAxisColumn);
+                $yAxisKeys = array_keys($yAxis);
+                $xQuery = $this->queryXAxis();
+                $xQuery->getSelectBase()->where($this->yAxisColumn, $yAxisKeys);
+                $xAxis = $this->getAxisKeys($xQuery, $this->xAxisColumn);
+                $xAxisKeys = array_keys($yAxis);
+            }
+        }
+
+        $pivotData = [];
+        $pivotHeader = [
+            'cols'  => $xAxis,
+            'rows'  => $yAxis
+        ];
+
+        if (! empty($xAxis) && ! empty($yAxis)) {
+            $this->baseQuery->filter(Filter::equal($this->gridcols[$this->xAxisColumn], $xAxisKeys));
+            $this->baseQuery->filter(Filter::equal($this->gridcols[$this->yAxisColumn], $yAxisKeys));
```

Comment:
Isn't it possible to use `$this->xAxisColumn` as column in the filter? Aliases are usually available for use in a WHERE.